### PR TITLE
Preprocessor support

### PR DESCRIPTION
**Pull request by sizmailov**
First I want to thank you for fast implementation of macro support! 

This PR pushes it a bit further. 

Changes: 
- `FortranMacrosContext` now has information about nested conditional directives 
- Before full macro support `#if` / `#elif` behaves as if the condition is `true` by default. It can be altered by trailing meta-comment `/* macro_eval: false */`
- a couple of minor fixes

### CODE REVIEW

File: src/main/kotlin/org/jetbrains/fortran/lang/preprocessor/FortranPreprocessingLexer.kt [hunk 1]
```diff
@@ -19,20 +20,45 @@ class FortranPreprocessingLexer : LookAheadLexer(FortranLexer(false)) {
         macrosContext.isDefined(identifier)
     }
     private val ifNotDefinedDecisionEvaluator = { name: CharSequence -> !ifDefinedDecisionEvaluator(name) }
+    private val ifDecisionEvaluator = { directive_content: CharSequence -> !directive_content.endsWith("/* macro_eval: false */") }
 
     override fun lookAhead(baseLexer: Lexer) {
+        val CONDITION_DIRECTIVES = TokenSet.create(IF_DIRECTIVE, IF_DEFINED_DIRECTIVE, IF_NOT_DEFINED_DIRECTIVE, ELSE_DIRECTIVE, ELIF_DIRECTIVE, ENDIF_DIRECTIVE)
```
Comment (sizmailov) [inline src/main/kotlin/org/jetbrains/fortran/lang/preprocessor/FortranPreprocessingLexer.kt]: This `CONDITION_DIRECTIVES` should be refactored out. 

I didn't get why, but several token sets are empty: `FortranTokenType.DIRECTIVES`,  `FortranTokenType.IF_DIRECTIVES`, `FortranTokenType.END_IF_DIRECTIVES`

**Comment by sizmailov** [comment]
The change is imperfect: 
 - non-compiled-region might be represented by number of PSI nodes
 - conditional directives are always colorful, even if they are in non-compiled-region

**Comment by sizmailov** [comment]
I run the patched version of plugin on 100K+ LOC code base and it does quite well. 

Of course problems with function-like and substitution macro are still here, but rest is almost flawless. 

@satamas Would it be possible to allow macro in more contexts without intrusive changes to grammar? Macro may appear virtually everywhere, would it be possible to treat them as white spaces?

Here are just a couple examples I see quite often which are not recognized by plugin correctly:

```fortran
subroutine foo(arg1, arg2 & 
               #ifdef MPI
                  , mpi_arg1, mpi_arg2 & 
               #endif 
              )
...
end subroutine foo
```


```fortran
if (condition &
    #ifdef MPI
        .or. mpi_condition &
    #endif 
   ) then 
... 
end if  
```


```fortran
! note the continuation &
if (condition) &  
#ifdef MPI
    call mpi_subroutine()
#else 
    call non_mpi_subroutine()
#endif 
```

The latter case is smelly and I would argue that it should be rewritten as `if .. then .. endif` (refactor intention suggested?), but it still should be recognized by plugin.

**Comment by satamas** [comment]
>  Would it be possible to allow macro in more contexts without intrusive changes to grammar?

I'm aware that parsing macros by adding them to grammar is a bad idea but I'm not sure that just ignoring them during parsing as it was before is a better idea. There must be some good way of supporting macros when you are using bnf based parser, but I don't know it. Will try to ask colleges to find one.

**Comment by sizmailov** [comment]
Found an issues: non-compiled statements are still "feel" each other and interact with compiled lines(!): 

image

PSI Viewer shows those regions as plain `PsiComment(conditionally_non_compiled_comment)`, i.e. there is no internal structure... 

Could it be that highlighting(?) engine uses another PSI without macro support?

**Comment by satamas** [comment]
> Would it be possible to allow macro in more contexts without intrusive changes to grammar? 
Macro may appear virtually everywhere, would it be possible to treat them as white spaces?

As it happens you were right and macro tokens should be just ignored by parser. I'll merge your pull request and continue with this brach, there are still a lot of problems and no tests

**Comment by sizmailov** [comment]
I cannot figure out why some of `TokenSet` variables of `FortranTokenType` are always empty, 
namely, `STRINGS`, `END_IF_DIRECTIVES`, `IF_DIRECTIVES`, `DIRECTIVES`

https://github.com/satamas/fortran-plugin/blob/92ad6838332dbe5c32f714185c7a02f0a049fd5f/src/main/java/org/jetbrains/fortran/lang/psi/FortranTokenType.kt#L53-L67

@satamas, can you help me with it ?

**Pull request merged**